Add caching to improve build time performance

adamus1red opened the pull request:
Uses the builtin `cache-from`/`cache-to` options from the `docker/buil-push-action` to write out the buildx cache.
Using the github `actions/cache` action the buildx cache is stored by github for future runs.

qdm12 commented:
Hey there, thanks for your PR!

However, having toyed around with buildx caching a few months ago, it always turned out to be slower than without cache, at least for Go projects where things are usually quick to download/test/build. What do you think? I can merge it and compare publish time by doing another commit on top too.

If that turns out to be faster, I can modify the `verify` job to use buildx and we can try adding caching there too.

qdm12 committed:
chore(ci): Docker base image layers caching
- Cache base image layers only
- Cache for verify job only
- Do not cache for publish job since:
  - most of the work is cross compiling
  - go.mod often change between publish jobs
  - prevent bugs introduced due to the cache (unlikely)

qdm12 committed:
try
